### PR TITLE
Add the ls-extension  - Metadata JSON to record converter

### PR DESCRIPTION
**Pull request by sachinira**
Add the language server extension, Metadata JSON to record converter. This will be used to generate Ballerina records from JSON objects which contains metadata related to CRM platforms such as Salesforce

Fixes 
https://github.com/ballerina-platform/ballerina-extended-library/issues/229

## Approach
The new Custom Record generator LS-Extension is intended to generate Ballerina records for the metadata obtained from CRM platforms. This will provide the capability to the developer to generate Ballerina records from metadata of CRM platforms

>Proposal
https://docs.google.com/document/d/1PL1VcT4vw1FEwn7gY2kVSWukBxjvSedh9MVtW7vIqIs/edit?usp=sharing
>Specification
https://docs.google.com/document/d/1RJWjafSo9F-o3KP59n-ZUA8mA5mCOMLXjYd-isyRkNo/edit?usp=sharing

Blank diagram - Page 1(1)

## Remarks
Related PRs
> https://github.com/wso2-enterprise/ballerina-low-code-editor/pull/1071
https://github.com/wso2-enterprise/ballerina-plugin-vscode/pull/446

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordClientCapabilities.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package io.ballerina.generators;
```
Comment (LakshanSS) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordClientCapabilities.java:1]: Missing licence header. Shall we check all the files and add it?

File: misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordClientCapabilities.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+import org.ballerinalang.langserver.commons.registration.BallerinaClientCapability;
+
+/**
+ * Client capabilities for the custom records service.
+ * The methods which are exposed to the client from the extension.
+ */
+public class MetadataToRecordClientCapabilities extends BallerinaClientCapability {
+    private boolean getResults;
+    private boolean getRemoteFunctionCalls;
```
Comment (RolandHewage) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordClientCapabilities.java:11]: Shall we use a better variable name here for the booleans.

File: misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordClientCapabilities.java [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        super(MetadataToRecordConstants.CAPABILITY_NAME);
+    }
+
+    public boolean isGetResults() {
+        return getResults;
+    }
+
+    public void setGetResults(boolean getResults) {
+        this.getResults = getResults;
+    }
```
Comment (RolandHewage) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordClientCapabilities.java:23]: Let's regenerate the getters and setters after adding a proper name for the boolean variables.

File: misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordConverterFactory.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import io.ballerina.generators.platform.SalesforceMetadataConverter;
+
+import static io.ballerina.generators.MetadataToRecordConstants.SALESFORCE;
+import static io.ballerina.generators.MetadataToRecordConstants.UNSUPPORTED_METADATA_SOURCE;
+
+/**
+ * Factory class for the creation of metadata converter.
+ */
+public class MetadataToRecordConverterFactory {
+    public MetadataConverter getConvert(String schemaSource) throws MetadataToRecordGeneratorException {
```
Comment (RolandHewage) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordConverterFactory.java:14]: Let's add documentation comments describing the functionality of individual public functions

File: misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordSchemaRequest.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ *  Request format for customRecordGenerator endpoint.
+ */
+public class MetadataToRecordSchemaRequest {
+    private String metadata;
+    private String schemaSource;
+    private boolean isClosed;
+    private boolean isRecordTypeDesc;
+
+    public MetadataToRecordSchemaRequest(String jsonString, String schemaSource, boolean isClosed) {
```
Comment (RolandHewage) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordSchemaRequest.java:12]: Let's add a more meaningful name for `jsonString` parameter like `metadataJsonString`

File: misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordServerCapabilities.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
+package io.ballerina.generators;
+
+import org.ballerinalang.langserver.commons.registration.BallerinaServerCapability;
+
+/**
+ * Server capabilities for the performance analyzer service.
+ */
+public class MetadataToRecordServerCapabilities extends BallerinaServerCapability {
+    private boolean getResults;
+    private boolean getRemoteFunctionCalls;
```
Comment (RolandHewage) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/MetadataToRecordServerCapabilities.java:10]: Let's give better variable names for these boolean variables

File: misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/exceptions/MetadataToRecordGeneratorException.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package io.ballerina.generators.exceptions;
+
+/**
+ *
+ */
```
Comment (RolandHewage) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/exceptions/MetadataToRecordGeneratorException.java:5]: Let's add the missing documentation comment

File: misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/platform/SalesforceMetadataConverter.java [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+                if (fieldData.isUpdateable() || fieldData.isCreateable()) {
+                    if (isNillable) {
+                        recordFieldNodeWithDefaultValue = createRecordFieldWithDefaultValueNode(null, null,
+                                createIdentifierToken(dataType +
+                                        createToken(QUESTION_MARK_TOKEN) + SF_WHITESPACE_TYPE),
+                                createIdentifierToken(fieldName),
+                                createToken(EQUAL_TOKEN),
+                                createSimpleNameReferenceNode(createIdentifierToken("()")),
+                                createToken(SEMICOLON_TOKEN));
+                        recordFieldList.add(recordFieldNodeWithDefaultValue);
```
Comment (RolandHewage) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/main/java/io/ballerina/generators/platform/SalesforceMetadataConverter.java:80]: Consider a scenario where a Salesforce object is created and will be updated in a next request. The default values may overwrite exisiting values if the user ignores the field.

File: misc/ls-extensions/modules/metadata-to-record-generator/src/test/java/io/ballerina/generators/MetadataToRecordServiceTest.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import io.ballerina.generators.platform.MetadataConverter;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+
```
Comment (RolandHewage) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/test/java/io/ballerina/generators/MetadataToRecordServiceTest.java:13]: Remove extra space

File: misc/ls-extensions/modules/metadata-to-record-generator/src/test/java/io/ballerina/generators/MetadataToRecordServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+//    @Test
+//    public void testSalesforceRecordTypeGeneration() throws IOException, MetadataToRecordGeneratorException {
+//        String file = "src/test/resources/json/opportunity.json";
+//        String json = new String(Files.readAllBytes(Paths.get(file)));
+//        MetadataToRecordConverterFactory factory = new MetadataToRecordConverterFactory();
+//        MetadataConverter source = factory.getConvert("Salesforce Schema");
+//        MetadataToRecordSchemaResponse response = source.convert(json);
+//    }
```
Comment (RolandHewage) [inline misc/ls-extensions/modules/metadata-to-record-generator/src/test/java/io/ballerina/generators/MetadataToRecordServiceTest.java:78]: Shall we remove the commented test

**Review by RolandHewage — COMMENTED** [review]
Added few comments.

**Comment by sachinira** [comment]
This PR is not needed because the suggested LS-Extension for Metadata to JSON converter development is discontinued development due to design decisions taken. Please refer the meeting notes here for more information
https://docs.google.com/document/d/1PL1VcT4vw1FEwn7gY2kVSWukBxjvSedh9MVtW7vIqIs/edit?usp=sharing

**Pull request closed without merging**